Show debug information about backend

abitrolly opened the pull request:
When reporting underlying backend (setuptools) issues, the (setuptools) tracker requirement is to mention exact version of backend being used. This information is, however, absent from the `build` output.

This PR shows backed spec being used, without version:

    $ touch pyproject.toml
    $ python -m build
    * Creating isolated environment: venv+pip... build:backend  setuptools.build_meta:__legacy__ build:python_executable  /tmp/build-env-098a9c2b/bin/python
    * Installing packages in isolated environment:
      - setuptools >= 40.8.0 ...

Getting version requires running custom code in venv created by `build` after backend is installed and it is not clear how to do that yet. The code should extract first component from the spec, and append `__version__` to it. At least that's the way of doing it for `setuptools`.

Fixes https://github.com/pypa/build/issues/959

abitrolly committed:
Show debug information about backend

When reporting underlying backend (setuptools) issues, the
(setuptools) tracker requirement is to mention exact version of
backend being used. This information is, however, absent from the
`build` output.

This commmit shows backed spec being used, without version:

    $ touch pyproject.toml
    $ python -m build
    * Creating isolated environment: venv+pip...
    build:backend  setuptools.build_meta:__legacy__
    build:python_executable  /tmp/build-env-098a9c2b/bin/python
    * Installing packages in isolated environment:
      - setuptools >= 40.8.0
    ...

Getting version requires running custom code in venv created by
`build` after backend is installed and it is not clear how to do
that yet. The code should extract first component from the spec,
and append `__version__` to it. At least that's the way of doing
it for `setuptools`.

Related to https://github.com/pypa/build/issues/959

layday reviewed (changes requested):
This should all probably be done in `__main__`, using `_ctx.log` with an `origin`.

abitrolly committed:
Use _ctx.log from __main__.py for debug output

    $ python -m build -v
    * Creating isolated environment: venv+pip...
    build:backend  setuptools.build_meta:__legacy__
    build:python_executable  /tmp/build-env-5yw5g814/bin/python
    ...

abitrolly committed:
Show build backend version in verbose mode

abitrolly committed:
Do not crash if backend info is unavailable

Also report backend lib name together with version

abitrolly committed:
Fix EncodingWarning: 'encoding' argument not specified

abitrolly committed:
Show backend version after build starts

Also strip newline from the version

abitrolly committed:
Show backend version in default verbosity

abitrolly committed:
Use 'debug' origin for diagnostic in verbose mode

abitrolly commented on src/build/_builder.py:
```diff
@@ -269,6 +271,18 @@ def prepare(
+        # setuptools.build_meta:__legacy__  -->  setuptools
+        lib = self._backend.split('.')[0]
+        script = f'import {lib}; print({lib}.__version__)'
+        cmd = [self.python_executable, '-c', script]
```
@layday `_compat` is unavailable inside venv. Is there any way to get it there?

abitrolly replied:
@layday I copied the `_compat.importlib` logic into script. That was the most straightforward solution.

layday replied:
`importlib_metadata` will not be available in the build env if nothing depends on it.

Spawning a subprocess is expensive - particularly on Windows - and we like to avoid doing it whenever we can.  It'd be better to pass the venv package folder path (purelib) in an as an arg to `version`, see e.g. https://github.com/pypa/build/blob/a40623b20de5d29c80b569283191d65e630f4fc0/src/build/env.py#L251C21-L251C25.

More importantly, this all doesn't need to happen inside `Builder.build`.  For one, programmatic users of the library will be paying the cost for information they've no interest in or have better ways of extracting.  We might also wanna print out dependency versions when performing other kinds of operations in the CLI, e.g. when we're extracting metadata from a project, which doesn't always involve a build.

Because of this and because the purelib path is not exposed by the `IsolatedEnv` interface, the best place to do all this work is the `__main__` module.

layday replied:
The official documentation is deficient. You can use [`distributions`](https://importlib-metadata.readthedocs.io/en/stable/api.html).

layday replied:
`purelib` wouldn't really be appropriate to use outside of a sys package-less venv as you'd be excluding packages elsewhere on `sys.path`.  I'm not sure if the `IsolatedEnv` interface needs to be genercised further - I think for now it'd probably be sufficient to add a method to the `DefaultIsolatedEnv` class to extract the versions of dependencies in it.  You can store the `purelib` in a private variable to reference from the new method.

abitrolly replied:
@layday `purelib` code doesn't work with `path`.

```sh
$ cd /tmp
$ python -m venv xxx
$ ls /tmp/xxx/lib/python3.13/site-packages
pip  pip-24.3.1.dist-info
$ python
```

```py
>>> from importlib.metadata import distributions
>>> next(iter(distributions(name="pip"))).version
'24.3.1'
>>> next(iter(distributions(name="pip", path="/tmp/xxx/lib/python3.13/site-packages"))).version
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    next(iter(distributions(name="pip", path="/tmp/xxx/lib/python3.13/site-packages"))).version
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
StopIteration
>>>
```

For my runs `_has_dependency` is only called with empty `kwargs`, so maybe the `purelib` codepath is dead.

What to do now?

layday replied:
The path is a `list[str]`.

abitrolly committed:
Use importlib to query backend version in build venv

abitrolly committed:
Use packages_distributions to get package name from module

abitrolly committed:
Print versions for all build requirements

It is easier and more informative than detecting package from
backend module name.

abitrolly commented:
Instead of detecting build backend module from its import path, then package name from the module name and then version from from package, just list installed versions of all build requirements.

````sh
* Building wheel...
    setuptools==80.9.0
```

abitrolly committed:
Use importlib to get venv package versions w/out subprocess

importlib discovery API takes optional path to site-packages,
so it is possible to query versios of packages inside venv
installation without activating it first.

To pass purelib_dir to builder (so that it can query the versions
of installed build backend packages for debugging), the path saved
in env backend (pip or uv class), and then env reference is passed
and saved in the builder.

abitrolly committed:
Add BaseEnv to query build dependency in system environment

Otherwise this API was only available in venv

abitrolly committed:
Add step origin to logging messages with *

This adds semantic to logging and decouples log messages

abitrolly committed:
Trying to fix tests by mocking query for installed packages

abitrolly committed:
Fix tests after letting default logger print all messages

Default logger used in tests ignored origin

abitrolly committed:
Fix projectbuilder log test

Tests use default logger that doesn't understand verbosity

abitrolly commented:
@gaborbernat rebased. Can we reopen that?

henryiii commented:
PRs can't be reopened after a rebase, you'll have to make a new one (from the same branch is fine).

abitrolly commented:
@henryiii well, now I know it. :grinning:
